src/fairseq2/utils/read_and_pickle_wrap_iterator.py

```diff
@@ -0,0 +1,48 @@
+from typing import Callable, Iterator, Tuple, TypeVar
+
+from typing_extensions import Self
+
+from fairseq2.data import DataPipelineBuilder, read_iterator
+
+T = TypeVar("T")
+
+
+class IteratorPickleWrapper(Iterator[T]):
+    def __init__(self, iterator_factory: Callable[[], Iterator[T]]) -> None:
```

Comment:
You might use a type alias for `Callable[[], Iterator[T]]` to make the code more readable:

```python
from typing_extensions import TypeAlias

IteratorFactory: TypeAlias = Callable[[], Iterator[T]]

class IteratorPickleWrapper(Iterator[T]):
  def __init__(self, factory: IteratorFactory[T]) -> None:
    ...
```